Latest Post Block (Post Content Support) iteration

AmartyaU opened the pull request:
## Description
<!-- Please describe what you have changed or added -->
Adds Post content support in #1594. @ajitbohra is working on all the other parts of the issue (#13698).

Work In Progress

### Screenshots

<img width="965" alt="Screen Shot 2019-03-30 at 10 51 01 PM" src="https://user-images.githubusercontent.com/44530193/55284122-4c9dd280-533e-11e9-9da9-eb0e9c8cc45c.png">

**Editor Side Display Excerpt**

<img width="1006" alt="Screen Shot 2019-03-30 at 10 45 33 PM" src="https://user-images.githubusercontent.com/44530193/55284098-ced9c700-533d-11e9-9768-c633a390b3fb.png">

**Rendered Display Excerpt**

<img width="881" alt="Screen Shot 2019-03-30 at 10 45 53 PM" src="https://user-images.githubusercontent.com/44530193/55284099-ced9c700-533d-11e9-8742-489d34bc34ae.png">

**Editor Side Display Full Post**

<img width="1000" alt="Screen Shot 2019-03-30 at 10 46 25 PM" src="https://user-images.githubusercontent.com/44530193/55284100-ced9c700-533d-11e9-8b5d-0bbe8bf42065.png">

**Rendered Display Full Post**

<img width="794" alt="Screen Shot 2019-03-30 at 10 46 46 PM" src="https://user-images.githubusercontent.com/44530193/55284101-ced9c700-533d-11e9-80fa-566e9720a7c7.png">


@gziolo @melchoyce You can track the progress on the issue here!

gziolo commented:
> I have not modified index.php because I am not sure if we are even using it, because I believe right now we are not using ServerSideRender to get the result from the index.php file and instead, we are just using the result from the latestPostsQuery to display everything on the frontend

Try to see the preview of the post with new options applied :) `ServerSideRender` has its own pros and cons. The benefit of using it is that you don't have to implement the same logic twice in both JavaScript and PHP, but the downside is that every time you change something to see the result you need to do network request. Latest Posts block is different as it has fully JS based `edit` implementation with immediate re-render after any change but the downside is that you need to re-implement PHP version as well.

gziolo commented on package-lock.json:
```diff
@@ -4753,8 +4753,7 @@
 							"bundled": true,
-							"dev": true,
-							"optional": true
+							"dev": true
```
That's surprising. Ideally, there should be no difference in `package-lock.json` file. Which version of npm do you use?

AmartyaU replied:
I am using 6.4.1

gziolo replied:
Can you remove changes to `package-lock.json` from this PR? You can copy the original lock file from `master` branch.

gziolo commented on packages/block-library/src/latest-posts/edit.js:
```diff
@@ -20,16 +16,18 @@ import {
 	BlockAlignmentToolbar,
 	BlockControls,
-} from '@wordpress/block-editor';
+} from '@wordpress/editor';
```
There are some ongoing code reorganizations and those components were moved to the new package. They will continue to work from `@wordpress/editor` (with maybe some warnings) but it should stay unchanged.

AmartyaU replied:
Thanks for the heads up! Should I change it back to @wordpress/block-editor too?

gziolo replied:
Yes.

AmartyaU commented:
I have completed most of the functionality, but I have a few design and development questions.

**Design Questions:**
1) In the latest post block, do we want to show the post the latest post block is in currently? 

I think this is redundant and may take up unnecessary space in 'Display Full Post' but may be useful in 'Display Post Excerpt', because it won't be repeating the same information already on the page in the latter case.

2) At the moment, if a post doesn't have an excerpt, I am not not displaying anything. Should I display the post content instead in this case?

I would prefer not displaying anything as an excerpt is inherently a summary and the first few lines of the post content wouldn't fit in in this category generally. 

3) In the 'Display Full Post', both on the editor and render side, I am displaying everything in a particular post except dynamic blocks (like RSS, latest post block). Is this fine?

**Development Question:**
1) In the PHP file, I am not able to escape HTML for the post_content before rendering it, because I need the display the full post HTML content with identical formatting. Is this safe? What would be a good workaround to ensure security with displaying HTML?

melchoyce commented:
> In the latest post block, do we want to show the post the latest post block is in currently?

My gut says "no," but I'm open to hearing other opinions.

> At the moment, if a post doesn't have an excerpt, I am not not displaying anything. Should I display the post content instead in this case?

We should show the post content truncated to whatever number of characters is being used for excerpts.

> In the 'Display Full Post', both on the editor and render side, I am displaying everything in a particular post except dynamic blocks (like RSS, latest post block). Is this fine?

Can you add some screenshots? Hard to visualize.

Thanks for all your work on this!

AmartyaU commented:
> > We should show the post content truncated to whatever number of characters is being used for excerpts.

Just updated the code to display truncated post content when excerpt is not present!

>> Can you add some screenshots? Hard to visualize.

For example, I added a RSS feed to the 'Bulleted Post' post
<img width="654" alt="Screen Shot 2019-04-02 at 11 27 24 AM" src="https://user-images.githubusercontent.com/44530193/55415168-93f0b280-553a-11e9-955f-4536516c0aa2.png">

But this RSS feed won't be displayed on either the editor or rendered side, because it's a dynamic block I am assuming 

<img width="718" alt="Screen Shot 2019-04-02 at 11 27 38 AM" src="https://user-images.githubusercontent.com/44530193/55415174-9521df80-553a-11e9-9eb2-b73c7ffd24e3.png">
<img width="851" alt="Screen Shot 2019-04-02 at 11 27 46 AM" src="https://user-images.githubusercontent.com/44530193/55415181-97843980-553a-11e9-9acf-92c6093026b4.png">

getdave commented:
> **Development Question:**
> 
> 1. In the PHP file, I am not able to escape HTML for the post_content before rendering it, because I need the display the full post HTML content with identical formatting. Is this safe? What would be a good workaround to ensure security with displaying HTML?

I think you might want [`wp_kses_post`](https://codex.wordpress.org/Function_Reference/wp_kses_post).

getdave commented:
@youknowriad Technical question. I can imagine developers needing to amend the markup generated by this Block. For example you might need the author to be displayed underneath the post content rather than above it. Would the [`blocks.getSaveElement` filter](https://wordpress.org/gutenberg/handbook/designers-developers/developers/filters/block-filters/#blocks-getsaveelement) be suitable for this purpose? Do you have any better suggestions?

I'm conscious that if a theme has a particular `loop()` setup, then the Theme author is going to need to write it once for the Theme PHP and again for this Block.

getdave commented on packages/block-library/src/latest-posts/index.php:
```diff
@@ -48,6 +50,24 @@ function render_block_core_latest_posts( $attributes ) {
+				 && isset( $attributes['displayPostContentRadio'] ) && $attributes['displayPostContentRadio'] == 'f' ) {
+			$list_items_markup .= sprintf(
+				'<div class="wp-block-latest-posts__post-full-content">%1$s</div>',
+				html_entity_decode( $post->post_content, ENT_QUOTES, get_option( 'blog_charset' ) )
```
`wp_kses_post()`?

youknowriad commented:
> blocks.getSaveElement filter 

This filter is in general not good to extend the block's output since it will create block invalidation issues. I tend to recommend the https://developer.wordpress.org/reference/hooks/render_block/ server side hook when it comes to this kind of usecase.

AmartyaU commented:
@getdave Thank you for testing the block out! Sorry, I should have clarified in the main post that I am only working on the Display Post Content part of the issue #1594. @ajitbohra is working on all the other parts of the issue #13698. 

> I think you might want [`wp_kses_post`](https://codex.wordpress.org/Function_Reference/wp_kses_post).

Thank you! This was exactly what I was looking for.

> It then goes on to suggest how to add one manually. I think we should have an option to add/remove a Read more link?

On this note, why doesn't the RSS feed have such an option, because I didn't see any when I was using it.
Just to clarify, do you mean adding an add/remove link on the user editor side?

> * Sticky posts - currently there is no option to respect [Sticky Posts](https://codex.wordpress.org/Sticky_Posts). This may well be intentional but it feels like it could be a useful feature. @melchoyce what say you?

This definitely could be useful for the users.

> * Default list styling in `list` layout mode - I've noticed the editor shows the default list bullet points when in list mode. It might make the Block look neater if these bullets were removed and the list "reset".

This is strange. I don't see any bullets in the default list mode, as you can see from my screenshots attached above.

getdave commented:
> Thank you for testing the block out! Sorry, I should have clarified in the main post that I am only working on the Display Post Content part of the issue #1594. @ajitbohra is working on all the other parts of the issue #13698.

No problem. I recommend you make this very clear both in this PR and in the core Issue. 

> On this note, why doesn't the RSS feed have such an option, because I didn't see any when I was using it. 

I don't know the answer to that. 

> Just to clarify, do you mean adding an add/remove 'Read More' link on the user editor side?

If the `Read more` is rendered on the server side (dynamic output) then we're probably all good. that said it would be nice to ensure the Editor reflects this if possible.

> This is strange. I don't see any bullets in the default list mode, as you can see from my screenshots attached above.

Odd. I will check next time I review the Block progress.


## Question

Are you actively working on this right now? Let me know if I can help you get this over the line.

mapk commented:
@getdave your feedback here was spot on. Thanks! @melchoyce asked me to take a look at the design questions while she's away, so here's a few more comments.

> Author meta - unless @melchoyce disagrees, I feel this should be styled so as to be visually consistent with the date in the meta.

Yes, let's style the author meta to match the date styling.

> Sticky posts - currently there is no option to respect Sticky Posts. This may well be intentional but it feels like it could be a useful feature. @melchoyce what say you?

It would be great to include an option to keep sticky posts at the top. While it's a great feature, I don't want it to be a blocker on this PR.

> Should we protect people from themselves and disable grid mode if the full post content is to be displayed?

My suggestion is to not worry about this. Yes, it makes for a bad layout, but I also don't want to overfill this with too many toggles and hidden modes.

> I've noticed the editor shows the default list bullet points when in list mode.

I can see these too. I'm using the Gutenberg Starter theme, so not sure if that makes a difference.

---

In this last one, I see two different alignment settings in the toolbar. One that shows all the alignments, and the other in the form of a dropdown.

The other part displays the layout, but I've selected "grid" and it isn't reflected in the toolbar option.

![toolbar](https://user-images.githubusercontent.com/617986/55600006-23e75580-570f-11e9-8b9d-14fea88f8439.png)

AmartyaU commented:
@mapk Thank you for your feedback! (though some of them may be more relevant for the #13698 PR)

>It would be great to include an option to keep sticky posts at the top. While it's a great feature, I don't want it to be a blocker on this PR.

Okay, I will work on this after I get all the other things done then.

>My suggestion is to not worry about this. Yes, it makes for a bad layout, but I also don't want to overfill this with too many toggles and hidden modes.

I am just going to leave this part as it is for now. One other simple thing we could do is not show the Display Full Post toggle when grid mode is on.

>I can see these too. I'm using the Gutenberg Starter theme, so not sure if that makes a difference.

I just saw there are li tags in the code. Do you want me to remove the bullets I the list mode?

<br> <br> 

> 1. In the latest post block, do we want to show the post the latest post block is in currently?

What are your views on the above? Mel suggested that we shouldn't display the current post.
 
> 3. In the 'Display Full Post', both on the editor and render side, I am displaying everything in a particular post except dynamic blocks (like RSS, latest post block). Is this fine?

What do you think about this? I also included screenshots in one of my comments above to explain this further.

getdave commented:
I agree with everything @mapk says. His comments supersede mine. I particularly agree with the points about not overcomplicating this with too many options or "nice to haves". Let's get the basics working first like @mapk says.

> I'm using the Gutenberg Starter theme, so not sure if that makes a difference.

It does indeed make a difference. @AmartyaU needs to be testing the Block against a Theme which does not enqueue Editor styles. 

@AmartyaU Please can you install and activate https://github.com/WordPress/gutenberg-starter-theme ?

Any _Theme specific_ overrides should be apply to the Theme's own Editor styles and should not be part of Gutenberg Core.

> What are your views on the above? Mel suggested that we shouldn't display the current post.

Thinking about this, outside the context of the page content there might be occasions when this will be required. However in the interests of keeping this PR focused, I'd say we don't look to include any controls around this. For the default setting I'd vote for checking how existing core functions behave (eg: https://codex.wordpress.org/Function_Reference/wp_get_recent_posts) and following that although I'm open to be contradicted by @mapk (or @melchoyce).


Thanks for your continued work here. It's great!

AmartyaU commented:
I agree. I am just going to focus on the basic functionality for now. Just made a commit adding the 'Read More' link on both the editor and server side!

**Editor Side**

<img width="1002" alt="Screen Shot 2019-04-05 at 9 49 47 AM" src="https://user-images.githubusercontent.com/44530193/55632474-64cf8080-5788-11e9-85a4-7a264613e5d9.png">

**Server Side**

<img width="886" alt="Screen Shot 2019-04-05 at 9 50 07 AM" src="https://user-images.githubusercontent.com/44530193/55632466-60a36300-5788-11e9-94b1-78e6dda003cd.png">

mapk commented:
>> In the latest post block, do we want to show the post the latest post block is in currently?
> 
> What are your views on the above? Mel suggested that we shouldn't display the current post.

I agree with Mel on this. Don't show the post that is currently being viewed. But @getdave brings up a good point about checking to see how Core currently does this, and follow that example. Can you look into that @AmartyaU? 


>> In the 'Display Full Post', both on the editor and render side, I am displaying everything in a particular post except dynamic blocks (like RSS, latest post block). Is this fine?
> 
> What do you think about this? I also included screenshots in one of my comments above to explain this further.

That seems fine to me.

AmartyaU commented:
@getdave @mapk I downloaded the Gutenberg starter theme and I can see the bullets now. Would you recommend taking them out in the list view?

>I agree with Mel on this. Don't show the post that is currently being viewed. But @getdave brings up a good point about checking to see how Core currently does this, and follow that example. Can you look into that @AmartyaU?

As far as I can see, the core displays all posts without filtering out the current post, so I guess we can go with that for now. 

I think we are mostly done with this PR!

getdave commented:
> @getdave @mapk I downloaded the Gutenberg starter theme and I can see the bullets now. Would you recommend taking them out in the list view?

I'm going to defer to @mapk and/or @melchoyce on this. My gut says it should look as close to the frontend as possible. I can't imagine many themes would countenance having list bullets show with content displayed in this format, but I could be wrong.

>  the core displays all posts without filtering out the current post, so I guess we can go with that for now.

Probably. It's a UX question really. It should be trivial to add a toggle for this in the future. @mapk what say you?

> I think we are mostly done with this PR!

Yes you've made great progress. I've tagged in for a Technical Review. If you don't get any joy from the Core team you can join the Core Editor Slack chat meeting on Wednesday, 3rd April 2019, 14:00 UTC.

gziolo commented:
There are still some PHP code styles issues to fix:

https://travis-ci.com/WordPress/gutenberg/jobs/190675748#L1009

![Screen Shot 2019-04-08 at 10 56 31](https://user-images.githubusercontent.com/699132/55711468-fccda400-59ec-11e9-99c2-51713d9e1942.png)

AmartyaU commented:
@gziolo Yeah, I noticed them. I was going to fix them at the end, so I am doing them now!

Commit:
Ran npm run test-e2e -- -u

gziolo commented:
@AmartyaU - something went wrong when you merged changes from the `master` branch in this commit: https://github.com/WordPress/gutenberg/pull/14627/commits/7cde3acc66e07f5715ada4b36e6bc17de229ad9c

It makes it almost impossible to review code changes. Can you try to fix it?

getdave commented:
> @AmartyaU - something went wrong when you merged changes from the `master` branch in this commit: [7cde3ac](https://github.com/WordPress/gutenberg/commit/7cde3acc66e07f5715ada4b36e6bc17de229ad9c)
> 
> It makes it almost impossible to review code changes. Can you try to fix it?

Thoughts as to what _could_ have gone wrong here:

1. You merged from [your repo's copy of `master`](https://github.com/AmartyaU/gutenberg) without first updating your `master` from [upstream `master`](https://github.com/WordPress/gutenberg) 
2. You use a `merge` rather than a `rebase` 

```
$ git checkout master
$ git pull upstream master && git push origin master
```

...then:

```
$ git checkout %%YOUR_BRANCH_NAME%%
$ git rebase master
```

getdave commented on packages/block-library/src/latest-posts/edit.js:
```diff
@@ -116,21 +114,20 @@ class LatestPostsEdit extends Component {
+							{ label: 'Excerpt', value: 'e' },
+							{ label: 'Full Post', value: 'f' },
+						] }
+						onChange={ ( value ) => this.updateAttribute( 'displayPostContentRadio', value ) }
```
**Minor**: currently you are creating a new anonymous function every time the render function runs.

Instead of creating a new anonymous function here could you [partially apply](https://medium.com/javascript-scene/curry-or-partial-application-8150044c78b8) the `displayPostContentRadio` argument, thereby allowing your handling function to be awaiting only 1 argument (ie: `value`). You could then do:

`onChange={ this.updateDisplayPostContentAttr }`

...thereby avoiding the anonymous function which should help performance wise.

AmartyaU replied:
I will look into this!

AmartyaU replied:
I am going to scrap the updateAttribute function and make all the changes in place as @draganescu suggested below

getdave commented on packages/block-library/src/latest-posts/index.php:
```diff
@@ -99,6 +99,14 @@ function register_block_core_latest_posts() {
+				),
+				'excerptLength' => array(
+					'type'    => 'number',
+					'default' => 55,
```
Is this `55` an arbitrary setting or does it mirror a default already in WP Core?

AmartyaU replied:
It mirrors other blocks like RSS that have an excerpt slider

getdave reviewed (commented):
I would review more changes but the merge is making this difficult as there are `358` files that have changed.

draganescu commented on packages/block-library/src/latest-posts/edit.js:
```diff
@@ -72,47 +70,97 @@ class LatestPostsEdit extends Component {
 
-	toggleDisplayPostDate() {
-		const { displayPostDate } = this.props.attributes;
+	updateAttribute( name, value ) {
```
@AmartyaU do we need this generic setter? It feels much better to have either the attribute set with an arrow functioni in the rendered component, or a function named specifically for what it does. I understand the need for DRY but sometimes this creates new API by mistake (e.g. we have setAttributes in the block API and it does not need an updateAttribute)

e.g. instead of 
`( value ) => this.updateAttribute( 'displayPostDate', value )` just do like you already do in other places `onChange={ ( value ) => setAttributes( { displayPostDate: value } ) }` and simply get rid of `updateAttribute` as it is useless.

AmartyaU replied:
I was using updateAttribute because it was left over in the code when I saw it, so I thought it might be a coding standard and was thus using it. I also prefer doing it in place so will make the changes.

AmartyaU replied:
Updated!

draganescu commented on packages/block-library/src/latest-posts/edit.js:
```diff
@@ -72,47 +70,97 @@ class LatestPostsEdit extends Component {
 		const { attributes, setAttributes, latestPosts } = this.props;
 		const { categoriesList } = this.state;
-		const { displayPostDate, align, postLayout, columns, order, orderBy, categories, postsToShow } = attributes;
+		const { displayFullPost, displayPostContent, displayPostFeaturedImage, displayPostDate, displayPostAuthor, align, postLayout, columns, order, orderBy, categories, postsToShow, wordsToShow } = attributes;
```
can we rename `wordsToShow` to `wordCount` or similar, indicating that it sets the number of words that will be displayed? named like this it suggests that we decide what/which wordsToShow.
on the same note `postsToShow ` has the same issue :)

AmartyaU replied:
Sure, will do!

AmartyaU replied:
I had actually already changed wordsToShow to excerptLength in the later commits which I think is even more descriptive

draganescu commented on packages/block-library/src/latest-posts/edit.js:
```diff
@@ -72,47 +70,97 @@ class LatestPostsEdit extends Component {
+					/>
+					{ displayPostContent &&
+					<RadioControl
+        		label="Show:"
```
is there an indentation problem here or is Github's problem?

AmartyaU replied:
I m pretty sure that's a Github problem. It looks fine in my editor and didn't raise any js lint errors when I was testing

draganescu commented on packages/block-library/src/latest-posts/edit.js:
```diff
@@ -72,47 +70,97 @@ class LatestPostsEdit extends Component {
+        		onChange={ ( value ) => this.updateAttribute( 'displayFullPost', value ) }
+    			/>
+					}
+					{ displayPostContent && displayFullPost=='e' &&
```
Usually strict equality is safer: `displayFullPost==='e'`

AmartyaU replied:
My bad, shouldn't have used this.

AmartyaU replied:
Actually I had already changes this in the latest commit

draganescu commented on packages/block-library/src/latest-posts/edit.js:
```diff
@@ -72,47 +70,97 @@ class LatestPostsEdit extends Component {
+        		selected={ displayFullPost }
+        		options={ [
+								{ label: 'Excerpt', value: 'e' },
+            		{ label: 'Full Post', value: 'f' },
```
Could we rename the values to be full explicit words? 
```
{ label: 'Excerpt', value: 'excerpt' },
{ label: 'Full Post', value: 'full' },
```
Is is much easier to follow later when you check for value.

AmartyaU replied:
I will rename these to full_post and excerpt then

draganescu reviewed (changes requested):
Awesome work @AmartyaU ! I left a few comments on the `6ddb5a5d` commit.

AmartyaU commented:
@gziolo @getdave Sorry, I did merge my branch. I will revert the changes and rebase later today.

Commit:
Revert "Ran npm run test-e2e -- -u"

This reverts commit 98e9b739b6a1aff6d7357dfef9858959da391ad1.

mapk commented:
> My gut says it should look as close to the frontend as possible.

^ regarding the bullets. I agree. Let's drop the bullets from the list view.



> the core displays all posts without filtering out the current post, so I guess we can go with that for now.
> 
> > Probably. It's a UX question really. It should be trivial to add a toggle for this in the future. @mapk what say you?

If that's what Core is doing now, let's keep that pattern.

gziolo commented on packages/block-library/src/latest-posts/edit.js:
```diff
@@ -203,6 +282,7 @@ export default withSelect( ( select, props ) => {
 		order,
 		orderby: orderBy,
 		per_page: postsToShow,
+		_embed: 1,
```
Out of curiosity, why `_embed` is prefixed with an underscore? Was it intended to be private?

AmartyaU replied:
Honestly I have no idea. I found it like this and was wondering about this too

draganescu replied:
_I think_ it is a rest/hateoas/hal convention to have parts of the response which are not part of the normal endpoint data prefixed with underscore (_links, _embedded) so agent enumerating props can skip these easily.

gziolo reviewed (commented):
Thanks for fixing the issue with merge with `master` branch. It's now much easier to review. It looks like you are on track and it's close to being ready. Great work addressing feedback so far 👍

draganescu commented on packages/block-library/src/latest-posts/edit.js:
```diff
@@ -20,19 +16,22 @@ import {
 
 /**
- * Module Constants
+ *  Module Constants
```
apparently there is an extra space here :)

AmartyaU replied:
Haha don't even remember putting it in there

draganescu commented on packages/block-library/src/latest-posts/edit.js:
```diff
@@ -168,27 +224,50 @@ class LatestPostsEdit extends Component {
+							{ displayPostFeaturedImage && post._embedded[ 'wp:featuredmedia' ] &&
+								<img
+									alt={ post._embedded[ 'wp:featuredmedia' ][ 0 ].title.rendered }
+									src={ post._embedded[ 'wp:featuredmedia' ][ 0 ].source_url }
```
@AmartyaU This can be skipped, but I've noticed that the style, in general, in rendering markup in React is to avoid such long access variables (`post._embedded[ 'wp:featuredmedia' ][ 0 ].title.rendered`)  and declare them upper in the code. It allows you to:

- use Object/Array [destructuring](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment) and have a simple short variable to use in the markup
- add a snippet of comment why on we have to use the magic zero to get the feature image

Because you are in a map() and it looks very nice maybe a helper function that handles the above could do it?

AmartyaU replied:
@draganescu As mentioned in the PR description, I am only working on the 'display post content' section of the latest post block. The part you mentioned was completed by @ajitbohra in #13698.

On this note, @gziolo @getdave should I remove the parts of #13698 from my PR or submit my PR with those other parts included?

gziolo replied:
Yes, you can remove them if they are covered in #13698.

AmartyaU replied:
Yeah, I still need to revise. I was having some trouble debasing so am figuring that out

gziolo commented on packages/block-library/src/latest-posts/edit.js:
```diff
@@ -20,19 +16,22 @@ import {
 
 /**
- * Module Constants
+ *  Module Constants
```
```suggestion
 * Module Constants
```

gziolo commented on package-lock.json:
```diff
@@ -1,6 +1,6 @@
 {
 	"name": "gutenberg",
-	"version": "5.3.0",
+	"version": "5.4.0",
```
It seems like your branch isn't aligned with the latest `master` changes 🤷‍♂️

AmartyaU replied:
Yeah, I still need to rebase. I was having some trouble rebasing so am figuring that out

AmartyaU replied:
Just rebased! The version should be updated now

Commit:
Rename postsToShow to postCount

Commit:
Rebased into master branch

Commit:
Remove redundant calls to updateAttribute and make changes in place under onChange

StaggerLeee commented:
Please move away from showing full content in Blog/Arhive page. Each time I see one of those primitive and old (and some new) Blogs I went crazy.

What is the point of showing the same content (in the character the same) in Archive page and in Post view Page.

draganescu commented:
There seems to be some missing margin in grid mode?

<img width="654" alt="Screenshot 2019-04-16 at 08 38 04" src="https://user-images.githubusercontent.com/107534/56184582-39b81d00-6023-11e9-84f3-38827e694e4a.png">

draganescu reviewed (changes requested):
Upon local testing there was some CSS misfire when in grid mode. Other than that it all seemed to work OK!

Commit:
Fix grid view CSS style error

AmartyaU commented:
> Upon local testing there was some CSS misfire when in grid mode. Other than that it all seemed to work OK!

I tried to fix the CSS error. Can you check one more time?

Commit:
Fixed editor 55 words error for post content case

AmartyaU commented:
Also there was one other bug in the code. In the editor side, when there is no excerpt for a post, only 55 words of the post content were displayed and [...] was added after that. The server side was working fine as expected. This was because the query I was using in edit.js returned the value in this format for this particular case. I have fixed this error now!

In the below example, Testing has no excerpt and Latest has an except. 

**Before**
<img width="1017" alt="Screen Shot 2019-04-21 at 8 17 05 AM" src="https://user-images.githubusercontent.com/44530193/56469922-eae8f980-640d-11e9-9fbc-82abd7a86c29.png">

**After**

<img width="1027" alt="Screen Shot 2019-04-21 at 8 19 56 AM" src="https://user-images.githubusercontent.com/44530193/56469955-42876500-640e-11e9-9801-2330502bd1d8.png">

draganescu commented:
Hi @AmartyaU the column spacing is fixed, however I found some new issues:

1. The read more link is floated when the post contains a gallery:

<img width="565" alt="Screenshot 2019-04-22 at 17 44 37" src="https://user-images.githubusercontent.com/107534/56506405-127fb500-6527-11e9-8259-1e4a194e29dc.png">

This causes even worse problems when the gallery is in the middle column:

<img width="581" alt="Screenshot 2019-04-22 at 17 45 13" src="https://user-images.githubusercontent.com/107534/56506420-20cdd100-6527-11e9-98f0-f125e270299f.png">

2. We should strip the tags and formatting of the excerpt so that we can be sure of the result, not like in this case the bullets ruin the layout:

<img width="516" alt="Screenshot 2019-04-22 at 17 47 05" src="https://user-images.githubusercontent.com/107534/56506469-3fcc6300-6527-11e9-86cd-680e7e47d72d.png">

AmartyaU commented:
> _I think we shouldn't display galleries at all, and strip tags would help in this problem too as all img tags will be removed._

I think that's the right thing to do! So basically I should strip all the tags on the editor side so that when post content is displayed in place of excerpt (when a post doesn't have an excerpt), it follows a proper format (without images)? I think server side already does this, so I can just replicate that behavior.

Commit:
Removed HTML tags from excerpt and only displaying text

AmartyaU commented:
I just added a commit that addresses proposes a solution, but I would appreciate it if you would take a look at the solution. Instead of using Regex, I used https://ctrlq.org/code/19813-convert-html-to-text in my solution and it looks to be working fine. The excerpt displayed on the editor is identical to that displayed on the server side in this solution as shown below:

**Full Post Content**

<img width="1057" alt="Screen Shot 2019-04-23 at 9 36 47 PM" src="https://user-images.githubusercontent.com/44530193/56626737-f7896f80-6610-11e9-8e75-333a4a30a74b.png">

**Excerpt on Editor Side** 


<img width="1022" alt="Screen Shot 2019-04-23 at 10 00 03 PM" src="https://user-images.githubusercontent.com/44530193/56627297-302a4880-6613-11e9-9746-4dd91d47f7fe.png">



**Excerpt on Server Side** 

<img width="698" alt="Screen Shot 2019-04-23 at 9 37 18 PM" src="https://user-images.githubusercontent.com/44530193/56627254-ff4a1380-6612-11e9-942a-5e1c7f416aeb.png">

getdave commented:
Great work on this.

> So basically I should strip all the tags on the editor side so that when post content is displayed in place of excerpt (when a post doesn't have an excerpt), it follows a proper format (without images)?

I'm concerned about this. Are we proposing stipping all blocks/tags from the "Full Content" version? 

I can envisage a use case where someone wants to display their full latest posts on a page. If all tags and blocks are stripped from the content then that use case will not work. 

It's far from ideal that the layout looks poor if this block is used within another Block or the post themselves contain complex Blocks. However, that seems like more of an edge case to me.

What about adding an option to strip tags/content rather than applying as a blanket default?

AmartyaU commented:
> I'm concerned about this. Are we proposing stipping all blocks/tags from the "Full Content" version?

No, we are not proposing to strip all blocks/tags from the "Full Content" version. The "Full Content" version will still display everything as before. We are just stripping the blocks/tags for the excerpt in the case when a post doesn't have an excerpt and has to use its post content. Since excerpts don't have formatting, we want to be consistent when a post uses its post content as its excerpt and that is the reason we are stripping all tags in that particular case!

draganescu commented on packages/block-library/src/latest-posts/edit.js:
```diff
@@ -163,13 +183,21 @@ class LatestPostsEdit extends Component {
+						if ( post.excerpt.raw === '' ) {
+							excerpt = post.content.raw;
+						}
+						const temp = document.createElement( 'div' );
```
`temp` is not a good name, let's change it to excerptElement. It then reads nice:

```
const excerptElement = document.createElement( 'div' );
excerptElement.innerHTML = excerpt;
excerpt = excerptElement.textContent || excerptElement.innerText || '';
```

draganescu commented:
@getdave indeed, the tags are stripped only from the excerpt, especially the one that is generated automagically from the post. Do you find any reason to allow them? The problem with tags in automatic excerpts is that you can mistakenly print unclosed tags and also some formatting or content (media) doesn't make any sense in an excerpt.

@AmartyaU great work! I have left only one comment. other than that we're good to go!

AmartyaU committed:
Changed constant name temp to excerptElement

getdave commented:
@AmartyaU One more thing I noticed is that when articles are displayed along with post content the markup isn't particularly accessible.

<img width="755" alt="Screen Shot 2019-04-30 at 13 54 18" src="https://user-images.githubusercontent.com/444434/56962928-96e9bd80-6b4f-11e9-8e8f-9f240645a6bd.png">

In this context my questions are:

1. Should the article be wrapped in an `<article>` tag as it is now a full piece of content and not just a title or title + excerpt. If feels like [syndicatable content in its own right](https://www.w3.org/TR/html5/semantics.html#the-article-element).
2. Given the above, should the title of the post be wrapped in a valid heading level tag? You will see from the screenshot above that if the article content contains headings (which it likely will) then the markup of the Block becomes quite confusing. It _might_ be less so if it were within an article.

These may be problems to be solved in a v2. If so I'm happy to raise a separate issue.

AmartyaU commented:
> These may be problems to be solved in a v2. If so I'm happy to raise a separate issue.

I think raising a separate issue about this would be best!

westonruter commented on packages/block-library/src/latest-posts/index.php:
```diff
@@ -89,37 +125,49 @@ function register_block_core_latest_posts() {
 					'type' => 'string',
 				),
-				'postsToShow'     => array(
+				'postCount'               => array(
```
@draganescu Isn't this change going to cause headaches for existing sites that have populated `postsToShow` (both in PHP and JS)? Doesn't there need to be some migration of the old `postsToShow` prop to the new `postShow` prop?

gziolo replied:
Good catch @westonruter. I think you are right, it should be updated to the old name to avoid all the dance with deprecations.

youknowriad commented:
Nice addition here. Thanks all.
